release/models/system/openconfig-system-grpc-tunnel.yang

```diff
@@ -0,0 +1,348 @@
+      default true;
+      description
+        "Use gRPC transport security (TLS).  Enabled by default.  This
+        leaf allows transport security to be disabled for use cases that
+        are not supported, such as lab testing.";
+    }
+
+    leaf certificate-id {
+      type string;
+      description
+        "Name of the certificate that is associated with client (e.g.
+        mutual authentication).  The certificate ID is provisioned
+        through other interfaces to the device, such as the gNOI
+        certificate management service.  If this leaf is provisioned and
+        transport-security is enabled, mutual authentication is implied.
+        If this leaf is not provisioned and transport-security is
+        enabled, then the client will only validate the server
+        certificate (one-way authentication).";
+    }
+
+    leaf source-address {
+      type oc-inet:ip-address;
+      description
+        "Source IPv4/IPv6 address to use in communications to the tunnel
+        server";
+    }
+
+    leaf network-instance {
+      type oc-netinst:network-instance-ref;
+      description
+        "The network instance to initiate outbound communications to the
+        tunnel server.  When unspecified, the DEFAULT_INSTANCE should be
+        used.";
+    }
+
+    leaf dscp {
+      type oc-inet:dscp;
+      description
+        "DSCP marking of packets generated by the gRPC tunnel client";
+    }
+
+    leaf-list target {
+      type union {
+        type target-type;
+        type string;
+      }
+      description
+        "The target TCP based protocols that should be enabled for
+        tunneling for the specified tunnel server instance session.  The
+        target may be a well defined target-type variant or a target
+        that is negotiated out of band.";
+    }
+
+  }
+
+  grouping grpc-tunnel-config {
+    description
+      "Global gRPC tunnel configuration and state parameters";
+
+    leaf retry-interval {
+      type uint64 {
+        range 1..max;
+      }
+      units seconds;
+      description
+        "The interval (in seconds) that a tunnel client attempts to
+        re-establish connection to a tunnel server after failure";
+    }
+  }
+
+  grouping target-options {
+    description
+      "Grouping for gRPC tunnel client target options";
+
+    container target-options {
+      description
+        "Enclosing container for tunnel client target options";
+
+      container config {
+        description
+          "Configuration data for tunnel client target options";
+
+        uses target-options-config;
+      }
+
+      container state {
+        config false;
+
+        description
+          "Operational state data for tunnel client target options";
+
+        uses target-options-config;
+      }
+    }
+  }
+
+  grouping target-options-config {
+    description
+      "Grouping for gRPC tunnel client target options configuration and
+      operational state";
+
+    leaf-list pattern {
```

Comment:
I may be missing something here, but I'm not clear what pattern this refers to. I didn't find any reference to these options (and what these various values mean) in the `tunnel.proto`. Is there some extension that's being used here, or something that I forgot about? thanks.

Reply:
I would put this in the bucket of a helpful addition for abstraction to the spec.

The `target-options` here are to give a customizable and variable based approach to populating the `target` string field in the protobuf message - a unique value to distinguish the tunnel connection

https://github.com/openconfig/grpctunnel/blob/master/proto/tunnel/tunnel.proto#L78

`target-options` brings in a `pattern`, `custom-string` and `delimeter`

The `pattern` defines an ordered leaf-list of of variables that can be derived from the system or user-defined to pack into the string.  For instance if `[ HOSTNAME VENDOR CUSTOM ]` are defined then the client would populate a `custom-string` of say `ID=2022` then protobuf `target` string might be constructed as `pr1.sea1:VendorA:ID=2022`

If the user populates the `delimeter` field (e.g. `\\` then it can override the implementations default delimeter) to produce a string such as `pr1.sea1\\VendorA\\ID=2022`

This also accounts for the simpler case where the user _always_ wants to define the `target` string statically thus the configuration would consist of `pattern = [ CUSTOM ]` with a `custom-string = "Whatever you wish goes here"`

Reply:
Essentially it provides the flexibility for both - an on-box dynamic generator based on runtime parameters or the ability to insert a completely user-defined string (or a mix of both).

Reducing this down to a string and taking away the complexity is also fine (might still be worthwhile to define supported characters, max string length, etc.. otherwise we'll see divergence among implementors)

@gwizdms - any additional comments here?